### ISSUE TITLE
event flag is ignored when the set flag is being used

### ISSUE THREAD

**Issue by YuviGold** [aquasecurity/tracee#1543]
This combination is necessary when a subset of event is requests only under a set of events.

for example, `open` under `fs` only. `open` exist in several categories.
```bash
➜  ~ docker run --name tracee --rm --pid=host --cgroupns=host --privileged -v /tmp/tracee:/tmp/tracee -v /lib/modules/:/lib/modules/:ro -v /usr/src:/usr/src:ro -it aquasec/tracee:latest trace --list | grep open
open                   [default syscalls fs fs_file_ops]        (const char* pathname, int flags, mode_t mode)
mq_open                [syscalls ipc ipc_msgq]                  (const char* name, int oflag, mode_t mode, struct mq_attr* attr)
openat                 [default syscalls fs fs_file_ops]        (int dirfd, const char* pathname, int flags, mode_t mode)
perf_event_open        [syscalls system]                        (struct perf_event_attr* attr, pid_t pid, int cpu, int group_fd, unsigned long flags)
open_by_handle_at      [syscalls fs fs_file_ops]                (int mount_fd, struct file_handle* handle, int flags)
open_tree              [syscalls]                               (int dfd, const char* filename, unsigned int flags)
fsopen                 [syscalls fs]                            (const char* fsname, unsigned int flags)
pidfd_open             [syscalls]                               (pid_t pid, unsigned int flags)
openat2                [default syscalls fs fs_file_ops]        (int dirfd, const char* pathname, struct open_how* how, size_t size)
security_file_open     [default lsm_hooks fs fs_file_ops]       (const char* pathname, int flags, dev_t dev, unsigned long inode, unsigned long ctime, int syscall)
```

Steps to reproduce:

only event
```bash
➜  ~ docker run --name tracee --rm --pid=host --cgroupns=host --privileged -v /tmp/tracee:/tmp/tracee -v /lib/modules/:/lib/modules/:ro -v /usr/src:/usr/src:ro -it aquasec/tracee:latest trace --trace 'e=open*' | head
TIME             UID    COMM             PID     TID     RET              EVENT                ARGS
09:17:03:648032  1000   Chrome_IOThread  3197    3280    139              openat               dirfd: -100, pathname: /dev/shm/.org.chromium.Chromium.zUOCVc, flags: O_RDWR|O_CREAT|O_EXCL, mode: 384
09:17:03:648080  1000   Chrome_IOThread  3197    3280    142              openat               dirfd: -100, pathname: /dev/shm/.org.chromium.Chromium.zUOCVc, flags: O_RDONLY, mode: 0
09:17:03:715665  1000   gnome-shell      2974    2974    29               openat               dirfd: -100, pathname: /proc/stat, flags: O_RDONLY, mode: 0
09:17:03:734358  0      systemd-journal  313     313     60               openat               dirfd: -100, pathname: /proc/26604/comm, flags: O_RDONLY|O_CLOEXEC, mode: 0
09:17:03:734422  0      systemd-journal  313     313     60               openat               dirfd: -100, pathname: /proc/26604/cmdline, flags: O_RDONLY|O_CLOEXEC, mode: 0
09:17:03:734459  0      systemd-journal  313     313     60               openat               dirfd: -100, pathname: /proc/26604/status, flags: O_RDONLY|O_CLOEXEC, mode: 0
09:17:03:734497  0      systemd-journal  313     313     60               openat               dirfd: -100, pathname: /proc/26604/attr/current, flags: O_RDONLY|O_CLOEXEC, mode: 0
09:17:03:734516  0      systemd-journal  313     313     60               openat               dirfd: -100, pathname: /proc/26604/sessionid, flags: O_RDONLY|O_CLOEXEC, mode: 0
09:17:03:734535  0      systemd-journal  313     313     60               openat               dirfd: -100, pathname: /proc/26604/loginuid, flags: O_RDONLY|O_CLOEXEC, mode: 0
```

with `s=fs` the `e=open` is ignored.
```bash
➜  ~ docker run --name tracee --rm --pid=host --cgroupns=host --privileged -v /tmp/tracee:/tmp/tracee -v /lib/modules/:/lib/modules/:ro -v /usr/src:/usr/src:ro -it aquasec/tracee:latest trace --trace s=fs --trace 'e=open*' | head
TIME             UID    COMM             PID     TID     RET              EVENT                ARGS
09:17:29:805254  1000   ThreadPoolServi  1       2       0                epoll_wait           epfd: 3, events: 0xDAC002C9980, maxevents: 32, timeout: 10
09:17:29:815358  1000   ThreadPoolServi  1       2       1                epoll_wait           epfd: 3, events: 0xDAC002C9980, maxevents: 32, timeout: -1
09:17:29:815472  1000   ThreadPoolServi  1       2       1                read                 fd: 15, buf: 0x7F39ACDE7427, count: 1
09:17:29:815480  1000   ThreadPoolServi  1       2       0                epoll_wait           epfd: 3, events: 0xDAC002C9980, maxevents: 32, timeout: 0
09:17:29:815416  1000   ThreadPoolSingl  1       2557    1                write                fd: 16, buf: 0x7F39A0F9184F, count: 1
09:17:29:818424  0      containerd-shim  165245  165293  101              read                 fd: 16, buf: 0xC000304000, count: 4096
09:17:29:818468  0      containerd-shim  165245  165293  101              write                fd: 18, buf: 0xC000304000, count: 101
09:17:29:818479  0      containerd-shim  165245  165293  -11              read                 fd: 16, buf: 0xC000304000, count: 4096
09:17:29:818487  0      containerd-shim  165245  165293  0                epoll_pwait          epfd: 4, events: 0xC000203800, maxevents: 128, timeout: 0, sigmask: 0x0, sigsetsize: 86951
```

**Comment by yanivagman** [comment]
On my evironment, things work as expected.
Event flag (in addition to set) is not ignored, and works as expected.
I used the command you provided, and greped for open:
```
docker run --name tracee --rm --pid=host --cgroupns=host --privileged -v /tmp/tracee:/tmp/tracee -v /lib/modules/:/lib/modules/:ro -v /usr/src:/usr/src:ro -it aquasec/tracee:latest trace --trace s=fs --trace 'e=open*' | grep open
10:40:36:976932  0      systemd-journal  334     334     -2               openat               dirfd: -100, pathname: /run/log/journal/9bf6fa841f9f4c17ae91f16ea24457cd/system.journal, flags: O_RDWR|O_NONBLOCK|O_CLOEXEC, mode: 0
10:40:37:038557  1000   Chrome_IOThread  5509    5537    0                security_file_open   pathname: /dev/shm/.com.google.Chrome.LXq06A, flags: O_RDWR|O_CREAT|O_EXCL|O_LARGEFILE, dev: 26, inode: 240781, ctime: 1647168037035031301
10:40:37:038536  1000   Chrome_IOThread  5509    5537    251              openat               dirfd: -100, pathname: /dev/shm/.com.google.Chrome.LXq06A, flags: O_RDWR|O_CREAT|O_EXCL, mode: 384
10:40:37:038572  1000   Chrome_IOThread  5509    5537    0                security_file_open   pathname: /dev/shm/.com.google.Chrome.LXq06A, flags: O_RDONLY|O_LARGEFILE, dev: 26, inode: 240781, ctime: 1647168037035031301
10:40:37:038569  1000   Chrome_IOThread  5509    5537    255              openat               dirfd: -100, pathname: /dev/shm/.com.google.Chrome.LXq06A, flags: O_RDONLY, mode: 0
```

Maybe the default behavior is not clear enough.
The following rules are used to determine which events will be traced:
1. If no events or sets where given, use the default set (simply called "default").
2. If either the event or set flags where given, the default set of events is ignored. In that case, the events to be traced is the combination of all events/sets given.
3. If a specific event from a set is not required, the user can choose event!=xxx, which will then ignore this event in the chosen set (which can be a combination of events and sets)

**Comment by YuviGold** [comment]
Hey @yanivagman thank you for your quick and elaborated response.

Once I `grep` it on my machine, the results are narrowed down of course. But it's not what I was expecting (since I explicitly expected only to get the `open*` events).

I believe that this behavior is not straightforward. Usually, when you provide a filter on top of another filter it narrows down the results from the available ones.
I'd expect the set that was specified to choose the specific set (as you showed instead of the `default`). While on top of that, once `events` are being used, to override it and create some kind of an `AND` query to get its sub results.
(I provided above all the `open` events while I wanted only the `open` events that are related to the `fs` set)

> the events to be traced is the combination of all events/sets given.

Basically, it seems that in your explanation, all queries are `OR`ed.
While the `help` message of `--trace help` explains the opposite.
```
Only events that match all trace expressions will be traced (trace flags are ANDed).
```

**Comment by yanivagman** [comment]
> 

Yes, you are right, it can be confusing.
We previously had events and sets as independant flags, but later combined them into the trace flag. We might want to reconsider the UX in the future (for sure before version 1.0). I agree we can improve it. For now we can just fix the docs.

**Comment by yanivagman** [comment]
The reason we have it as an OR and not as an AND is for a user to be able to specify several sets, or several events at once. Using AND, that won't be possible. I think that scenario is far more useful than the one you described (choosing a set, and filtering  events from it). In addition, the scenario you described can be easily achieved using the != operator (not as easy as if we had ANDed the sets and events, but yet possible)
